Integrate RenderListMarker with @counter-style

vitorroriz opened the pull request:
#### 2e241f9b533bf886556c2a94a0cb242297a9d000
<pre>
Integrate RenderListMarker with @counter-style
<a href="https://bugs.webkit.org/show_bug.cgi?id=253158">https://bugs.webkit.org/show_bug.cgi?id=253158</a>
rdar://102988393

Reviewed by Tim Nguyen.

This patch integrates RenderListMarker with
the parsed counter-style at-rules and fixes
some bugs catched during rendering.

Since we are wiring things up here, we can stop
skipping the related tests.

Some tests will still fail because we are missing:
- Support to shadow DOM scope.
- Integration with counters() function.
- Integration with symbols() function.
- Complex predefined symbols.
- Allowing redefinition of predefined symbols.
- Bug on wpt references regarding prefix/suffix fallback.
The missing items are linked to their issues in TestExpectations.

* LayoutTests/TestExpectations:
- Stop skipping the related tests. We are also tagging the tests
that we still fail due to missing feature/wpt bugs and we are
linking the relevant issues.

* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/counter-style-at-rule/support/ref-common.css:
(div, p):
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/counter-style-at-rule/support/test-common.css:
(li, p):
- Avoid fuzziness in the relevant tests. This change is also pushed to WPT (<a href="https://github.com/web-platform-tests/wpt/pull/38773)">https://github.com/web-platform-tests/wpt/pull/38773)</a>

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemCyclic const):
(WebCore::CSSCounterStyle::applyPadSymbols const):
- Fixing bugs in padding and cyclic-system.

* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::resolveExtendsReference):
- We were forgetting to pass the map for reference resolution
during the resolution of author rules, which would make the
user agent map to be used instead, by default.

* Source/WebCore/css/CSSProperties.json:
- Enabling custom-ident parsing in list-style-type.

* Source/WebCore/css/counterStyles.css:
(@counter-style cjk-decimal):
(@counter-style disclosure-open):
(@counter-style disclosure-closed):
- Update the symbols for disclosure open/closed.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleAdditiveSymbols):
- Fixing bug where we would not invalidate additive-symbol with missing symbol.
This would cause a crash during rendering.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout):
- listStyleStringValue change now requires layout, since it defines the name of the counter-style.
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::updateContent):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueListStyleType):
- Rendering Marker according to counter-style.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-type-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-type-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/list-style-type-expected.txt:
- We have new tests passing after fixing presentation hint mapping for LI and UL elements.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):

* Source/WebCore/html/HTMLLIElement.cpp:
(WebCore::HTMLLIElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLUListElement.cpp:
(WebCore::HTMLUListElement::collectPresentationalHintsForAttribute):
- Fixing presentational hint mapping for LI and UL elements.

Canonical link: <a href="https://commits.webkit.org/261135@main">https://commits.webkit.org/261135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f54742806aa11444b7313bc53435eef333ef086e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110694 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43256 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/2046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114661 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21184 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Running trigger") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10888 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1555 "layout-tests (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116436 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103009 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30619 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/44119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12958 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51575 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7714 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14895 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -1843,13 +1843,19 @@ void RenderListMarker::updateContent()
     auto type = style().listStyleType();
     switch (type) {
     // FIXME: handle CSSCounterStyle case rdar://102988393.
```
Can we remove this comment?

alanbaradlay commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -1843,13 +1843,19 @@ void RenderListMarker::updateContent()
+        auto text = makeString(counterStyle()->prefix(), counterStyle()->text(m_listItem->value()));
+        m_textWithSuffix = makeString(text, counterStyle()->suffix());
+        m_textWithoutSuffixLength = text.length();
+        m_textIsLeftToRightDirection = u_charDirection(text[0]) != U_RIGHT_TO_LEFT;
```
is it ok to just checking the first character? (or go with ubidi_getBaseDirection)

vitorroriz replied:
I've followed the pattern that is already done for all other ListStyleTypes (case default) and case String.

alanbaradlay replied:
sure, though that does not necessarily mean it is correct.

alanbaradlay replied:
oh look, there's a FIXME about this right there.
// FIXME: Depending on the string value, we may need the real bidi algorithm.

vitorroriz replied:
I see, but we don't have tests failing for now. Shouldn't we investigate this in another radar? I'm adding a new value to ListStyleType.

in `RenderListMarker::textRun()` there is a comment:

>     // Since the bidi algorithm doesn't run on this text, we instead reorder the characters here.
>     // We use u_charDirection to figure out if the marker text is RTL and assume the suffix matches the surrounding direction.

I've created rdar://106139180 so we can investigate it. I'll update the FIXME comment with the radar link after current pipeline finish running so it don't get reset.

alanbaradlay replied:
Perfect! Thanks!

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -8281,6 +8281,8 @@ RefPtr<CSSValue> consumeCounterStyleAdditiveSymbols(CSSParserTokenRange& range,
                 return nullptr;
         }
+        if (!symbol)
+            return nullptr;
```
Can we check for !symbol before checking for !integer and remove the !symbol check inside !integer?

vitorroriz replied:
the symbol and weight can appear in any order `[<integer [0,∞]> && <symbol>]#`

nt1m commented on Source/WebCore/html/HTMLLIElement.cpp:
```diff
@@ -77,8 +77,19 @@ void HTMLLIElement::collectPresentationalHintsForAttribute(const QualifiedName&
+            else if (valueLowerCase == "square"_s)
+                addPropertyToPresentationalHintStyle(style, CSSPropertyListStyleType, CSSValueSquare);
+            else if (valueLowerCase == "none"_s)
+            addPropertyToPresentationalHintStyle(style, CSSPropertyListStyleType, CSSValueNone);
```
please indent this

nt1m commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -41,9 +41,14 @@ namespace WebCore {
+    }
+
+    unsigned symbolIndex { 0 };
+    symbolIndex = static_cast<unsigned>(value > 0 ? value : value + amountOfSymbols);
```
```suggestion
    unsigned symbolIndex = static_cast<unsigned>(value > 0 ? value : value + amountOfSymbols);
```

A comment explaining this is to avoid substracting 1 from the minimum integer would be nice.

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -1840,16 +1837,29 @@ void RenderListMarker::updateContent()
+            m_textWithoutSuffixLength = text.length();
+            m_textIsLeftToRightDirection = u_charDirection(text[0]) != U_RIGHT_TO_LEFT;
+        } else
+            markerAsString();
```
This seems incorrect? If the counter-style doesn't exist, it should fallback to decimal I think.

vitorroriz committed:
Integrate RenderListMarker with @counter-style
https://bugs.webkit.org/show_bug.cgi?id=253158
rdar://102988393

Reviewed by Tim Nguyen.

This patch integrates RenderListMarker with
the parsed counter-style at-rules and fixes
some bugs catched during rendering.

Since we are wiring things up here, we can stop
skipping the related tests.

Some tests will still fail because we are missing:
- Support to shadow DOM scope.
- Integration with counters() function.
- Integration with symbols() function.
- Complex predefined symbols.
- Allowing redefinition of predefined symbols.
- Bug on wpt references regarding prefix/suffix fallback.
The missing items are linked to their issues in TestExpectations.

* LayoutTests/TestExpectations:
- Stop skipping the related tests. We are also tagging the tests
that we still fail due to missing feature/wpt bugs and we are
linking the relevant issues.

* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/counter-style-at-rule/support/ref-common.css:
(div, p):
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/counter-style-at-rule/support/test-common.css:
(li, p):
- Avoid fuzziness in the relevant tests. This change is also pushed to WPT (https://github.com/web-platform-tests/wpt/pull/38773)

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemCyclic const):
(WebCore::CSSCounterStyle::applyPadSymbols const):
- Fixing bugs in padding and cyclic-system.

* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::resolveExtendsReference):
- We were forgetting to pass the map for reference resolution
during the resolution of author rules, which would make the
user agent map to be used instead, by default.

* Source/WebCore/css/CSSProperties.json:
- Enabling custom-ident parsing in list-style-type.

* Source/WebCore/css/counterStyles.css:
(@counter-style cjk-decimal):
(@counter-style disclosure-open):
(@counter-style disclosure-closed):
- Update the symbols for disclosure open/closed.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleAdditiveSymbols):
- Fixing bug where we would not invalidate additive-symbol with missing symbol.
This would cause a crash during rendering.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout):
- listStyleStringValue change now requires layout, since it defines the name of the counter-style.
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::updateContent):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueListStyleType):
- Rendering Marker according to counter-style.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-type-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-type-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/list-style-type-expected.txt:
- We have new tests passing after fixing presentation hint mapping for LI and UL elements.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):

* Source/WebCore/html/HTMLLIElement.cpp:
(WebCore::HTMLLIElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLUListElement.cpp:
(WebCore::HTMLUListElement::collectPresentationalHintsForAttribute):
- Fixing presentational hint mapping for LI and UL elements.

Canonical link: https://commits.webkit.org/261135@main
